#                   Capstone Project 1: Employer Ratings Analysis

# Data Wrangling Steps:
The data has been acquired as a csv file from https://www.kaggle.com/petersunga/google-amazon-facebook-employee-reviews. It #was created by web scraping over 67K employee reviews for employers such as Amazon, Apple, Google, Facebook, Microsoft and #Netflix. The reviews were collected over a time span of 10 years (2009-2018).

# Import packages and read the csv file into a pandas DataFrame. 

In [1]:
import numpy as np
import pandas as pd
emp_df = pd.read_csv('employee_reviews.csv')

In [2]:
emp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67529 entries, 0 to 67528
Data columns (total 17 columns):
Unnamed: 0                    67529 non-null int64
company                       67529 non-null object
location                      67529 non-null object
dates                         67529 non-null object
job-title                     67529 non-null object
summary                       67409 non-null object
pros                          67529 non-null object
cons                          67529 non-null object
advice-to-mgmt                67232 non-null object
overall-ratings               67529 non-null float64
work-balance-stars            67529 non-null object
culture-values-stars          67529 non-null object
carrer-opportunities-stars    67529 non-null object
comp-benefit-stars            67529 non-null object
senior-mangemnet-stars        67529 non-null object
helpful-count                 67529 non-null int64
link                          67529 non-null object
dtypes: 

In [3]:
emp_df.head()

Unnamed: 0 company           location          dates  \
0           1  google               none   Dec 11, 2018   
1           2  google  Mountain View, CA   Jun 21, 2013   
2           3  google       New York, NY   May 10, 2014   
3           4  google  Mountain View, CA    Feb 8, 2015   
4           5  google    Los Angeles, CA   Jul 19, 2018   

                                  job-title  \
0     Current Employee - Anonymous Employee   
1         Former Employee - Program Manager   
2  Current Employee - Software Engineer III   
3     Current Employee - Anonymous Employee   
4       Former Employee - Software Engineer   

                                             summary  \
0                           Best Company to work for   
1  Moving at the speed of light, burn out is inev...   
2  Great balance between big-company security and...   
3  The best place I've worked and also the most d...   
4                    Unique, one of a kind dream job   

                                                pros  \
0                      People are smart and friendly   
1  1) Food, food, food. 15+ cafes on main campus ...   
2  * If you're a software engineer, you're among ...   
3  You can't find a more well-regarded company th...   
4  Google is a world of its own. At every other c...   

                                                cons  \
0                 Bureaucracy is slowing things down   
1  1) Work/life balance. What balance? All those ...   
2  * It *is* becoming larger, and with it comes g...   
3  I live in SF so the commute can take between 1...   
4  If you don't work in MTV (HQ), you will be giv...   

                                      advice-to-mgmt  overall-ratings  \
0                                               none              5.0   
1  1) Don't dismiss emotional intelligence and ad...              4.0   
2  Keep the focus on the user. Everything else wi...              5.0   
3  Keep on NOT micromanaging - that is a huge ben...              5.0   
4  Promote managers into management for their man...              5.0   

  work-balance-stars culture-values-stars carrer-opportunities-stars  \
0                4.0                  5.0                        5.0   
1                2.0                  3.0                        3.0   
2                5.0                  4.0                        5.0   
3                2.0                  5.0                        5.0   
4                5.0                  5.0                        5.0   

  comp-benefit-stars senior-mangemnet-stars  helpful-count  \
0                4.0                    5.0              0   
1                5.0                    3.0           2094   
2                5.0                    4.0            949   
3                4.0                    5.0            498   
4                5.0                    5.0             49   

                                                link  
0  https://www.glassdoor.com/Reviews/Google-Revie...  
1  https://www.glassdoor.com/Reviews/Google-Revie...  
2  https://www.glassdoor.com/Reviews/Google-Revie...  
3  https://www.glassdoor.com/Reviews/Google-Revie...  
4  https://www.glassdoor.com/Reviews/Google-Revie...

The dataset consists of 67529 observations with 17 columns. 

# Some of the columns are poorly labeled so they should be renamed.

In [4]:
emp_df = emp_df.rename(columns = {"job-title": "job_title", "advice-to-mgmt": "advice_to_mgmt", 
                                        "overall-ratings": "overall_ratings", "work-balance-stars": "work_balance_stars", 
                                        "culture-values-stars": "culture_values_stars", "helpful-count": "helpful_count",
                                        "carrer-opportunities-stars": "career_opportunities_stars",
                                        "comp-benefit-stars": "comp_benefit_stars", "senior-mangemnet-stars": "senior_management_stars"})

# Check for null values in all columns

In [5]:
emp_df.isnull().sum()

Unnamed: 0                      0
company                         0
location                        0
dates                           0
job_title                       0
summary                       120
pros                            0
cons                            0
advice_to_mgmt                297
overall_ratings                 0
work_balance_stars              0
culture_values_stars            0
career_opportunities_stars      0
comp_benefit_stars              0
senior_management_stars         0
helpful_count                   0
link                            0
dtype: int64

# Check for some unique values

In [6]:
emp_df['location'].unique()

array(['none', 'Mountain View, CA', 'New York, NY', ...,
       'Srinagar, Jammu and Kashmir (India)', 'Blackburn, England (UK)',
       'Pudong, Shanghai (China)'], dtype=object)

In [7]:
emp_df['work_balance_stars'].unique()

array(['4.0', '2.0', '5.0', 'none', '3.0', '1.0', '4.5', '2.5', '3.5',
       '1.5'], dtype=object)

Some columns have “none” as value. So they do not qualify as missing value, even though they are null. By replacing all “none” values with np.NaN, they can be identified as a null value.

# Replace 'none' with np.Nan

In [8]:
emp_df2 = emp_df.replace(to_replace = 'none', value = np.NaN)
emp_df2.isnull().sum()

Unnamed: 0                        0
company                           0
location                      25085
dates                             0
job_title                         0
summary                         129
pros                              0
cons                              1
advice_to_mgmt                29461
overall_ratings                   0
work_balance_stars             7160
culture_values_stars          13546
career_opportunities_stars     7108
comp_benefit_stars             7161
senior_management_stars        7775
helpful_count                     0
link                              0
dtype: int64

By replacing 'none' with np.NaN, we can see that several columns have missing values.

In [9]:
emp_df2.dtypes

Unnamed: 0                      int64
company                        object
location                       object
dates                          object
job_title                      object
summary                        object
pros                           object
cons                           object
advice_to_mgmt                 object
overall_ratings               float64
work_balance_stars             object
culture_values_stars           object
career_opportunities_stars     object
comp_benefit_stars             object
senior_management_stars        object
helpful_count                   int64
link                           object
dtype: object

The ratings columns (overall_ratings, work_balance_stars, culture_values_stars, career_opportunities_stars, comp_benefit_stars, senior_management_stars) are the most important variables in this project. In order to fill the missing values on them, they will be first converted from object datatype to float type. And then, the null values will be replaced with mean of the column.

# Convert object types to float and replace null values with mean of the column for better analysis

In [10]:
emp_df2['work_balance_stars'] = emp_df2['work_balance_stars'].astype(str).astype(float)

wbs_mean = emp_df2['work_balance_stars'].mean()

emp_df2['work_balance_stars'] = emp_df2['work_balance_stars'].fillna(value=wbs_mean)

In [11]:
emp_df2['culture_values_stars'] = emp_df2['culture_values_stars'].astype(str).astype(float)

cvs_mean = emp_df2['culture_values_stars'].mean()

emp_df2['culture_values_stars'] = emp_df2['culture_values_stars'].fillna(value=cvs_mean)

In [12]:
emp_df2['career_opportunities_stars'] = emp_df2['career_opportunities_stars'].astype(str).astype(float)

cos_mean = emp_df2['career_opportunities_stars'].mean()

emp_df2['career_opportunities_stars'] = emp_df2['career_opportunities_stars'].fillna(value=cos_mean)

In [13]:
emp_df2['comp_benefit_stars'] = emp_df2['comp_benefit_stars'].astype(str).astype(float)

cbs_mean = emp_df2['comp_benefit_stars'].mean()

emp_df2['comp_benefit_stars'] = emp_df2['comp_benefit_stars'].fillna(value=cbs_mean)

In [14]:
emp_df2['senior_management_stars'] = emp_df2['senior_management_stars'].astype(str).astype(float)

sms_mean = emp_df2['senior_management_stars'].mean()

emp_df2['senior_management_stars'] = emp_df2['senior_management_stars'].fillna(value=sms_mean)

# Sort the dataframe by company and fill the null values in location column using the ffill and bfill methods. 

In [15]:
emp_df_sort = emp_df2.sort_values('company')
emp_df_sort['location'] = emp_df_sort['location'].fillna(method = 'bfill')

In [16]:
emp_df_sort.isnull().sum()

Unnamed: 0                        0
company                           0
location                          1
dates                             0
job_title                         0
summary                         129
pros                              0
cons                              1
advice_to_mgmt                29461
overall_ratings                   0
work_balance_stars                0
culture_values_stars              0
career_opportunities_stars        0
comp_benefit_stars                0
senior_management_stars           0
helpful_count                     0
link                              0
dtype: int64

In [17]:
emp_df_sort[emp_df_sort['location'].isnull()]

Unnamed: 0  company location          dates  \
36647       36648  netflix      NaN   Apr 20, 2018   

                                  job_title        summary  \
36647  Former Employee - Anonymous Employee  Art et médias   

                                                    pros  \
36647  Apprendre à travailler sur pression. Développe...   

                                          cons  \
36647  Stimuler à faire beaucoup plus qu’avant   

                                advice_to_mgmt  overall_ratings  \
36647  Toujours donner l’opportunité au jeune.              5.0   

       work_balance_stars  culture_values_stars  career_opportunities_stars  \
36647            3.373254              3.784451                    3.634035   

       comp_benefit_stars  senior_management_stars  helpful_count  \
36647            3.942503                 3.322522              0   

                                                    link  
36647  https://www.glassdoor.com/Reviews/Netflix-Revi...

In [18]:
emp_df_sort['location'] = emp_df_sort['location'].fillna(method = 'ffill')
emp_df_sort[emp_df_sort['location'].isnull()]

Empty DataFrame
Columns: [Unnamed: 0, company, location, dates, job_title, summary, pros, cons, advice_to_mgmt, overall_ratings, work_balance_stars, culture_values_stars, career_opportunities_stars, comp_benefit_stars, senior_management_stars, helpful_count, link]
Index: []

# Check the values in dates column

In [19]:
emp_df_sort['dates'].isnull().sum()

0

# Convert the dates column from object to datetime

In [20]:
emp_df_sort.loc[emp_df_sort['dates'] == 'None']

Unnamed: 0 company     location dates  \
7819        7820  amazon  Phoenix, AZ  None   

                                            job_title  \
7819  Current Employee - Software Development Manager   

                       summary  \
7819  An Amazing Place to Work   

                                                   pros  \
7819  I've been at Amazon for a month now, and I've ...   

                                                   cons advice_to_mgmt  \
7819  No cons, so far - seriously. Like I said, I'm ...            NaN   

      overall_ratings  work_balance_stars  culture_values_stars  \
7819              5.0                 4.0                   5.0   

      career_opportunities_stars  comp_benefit_stars  senior_management_stars  \
7819                         5.0                 5.0                      5.0   

      helpful_count                                               link  
7819            580  https://www.glassdoor.com/Reviews/Amazon-Revie...

In [21]:
emp_df_sort['dates'] = pd.to_datetime(emp_df_sort['dates'], errors='coerce')

At the end of this process, the only variables that continue to have null values are summary and advice_to_mgmt but that can be ignored for our analysis.